frontend/src/composables/useShootEditor.js

```diff
@@ -0,0 +1,254 @@
+        extraKey.key = get(localStorageStore.editorShortcuts, [extraKey.key])
+      }
+    }
+    return extraKeys
+  }
+
+  const isReadOnly = computed(() => {
+    return isShootActionsDisabled.value || !authzStore.canPatchShoots
+  })
+
+  const isDarkMode = computed(() => {
+    return !!theme.global.current.value?.dark
+  })
+
+  const shootItem = computed(() => {
+    const value = unref(initialValue)
+    if (!value) {
+      return null
+    }
+    const {
+      kind,
+      apiVersion,
+      metadata,
+      spec,
+      status,
+    } = value
+    return {
+      kind,
+      apiVersion,
+      metadata: !showManagedFields.value && metadata?.managedFields
+        ? omit(metadata, 'managedFields')
+        : metadata,
+      spec,
+      status,
+    }
+  })
+
+  const filename = computed(() => {
+    const namespace = get(shootItem.value, ['metadata', 'namespace'])
+    if (!namespace) {
+      return get(options, ['filename'], 'unknown.yaml')
+    }
+    const name = get(shootItem.value, ['metadata', 'name'], 'unnamed')
+    const projectName = projectStore.projectNameByNamespace(namespace)
+    return `shoot--${projectName}--${name}.yaml`
+  })
+
+  const isShootActionsDisabled = computed(() => {
+    return get(shootItem.value, ['spec', 'purpose']) === 'infrastructure'
+  })
+
+  let cm = null
+
+  async function loadEditor (element) {
+    try {
+      const { useCodemirror } = await import('./useCodemirror')
+      cm = useCodemirror(element, {
+        ...options,
+        onDocumentChanged ({ modified, undoDepth, redoDepth }) {
+          touched.value = true
+          clean.value = modified
+          historySize.value = {
+            undo: undoDepth,
+            redo: redoDepth,
+          }
+        },
+        onTooltip ({ visible, ...rest }) {
+          helpTooltip.visible = visible
+          if (visible) {
+            Object.assign(helpTooltip, rest)
+          }
+        },
+        doc: yaml.dump(shootItem.value),
+        extraKeys: getExtraKeys(),
+        readOnly: isReadOnly.value,
+        darkMode: isDarkMode.value,
+      })
+    } catch (err) {
+      logger.error('Failed to create codemirror instance: %s', err.message)
+      logger.error(err.stack)
+    }
+  }
+
+  function destroyEditor () {
+    cm?.destroy()
+  }
+
+  function getEditorValue () {
+    const value = cm?.getDocValue()
+    return value
+      ? yaml.load(value)
+      : null
+  }
+
+  function setEditorValue (value) {
+    if (value) {
+      cm?.setDocValue(yaml.dump(value))
+    }
+  }
+
+  function resetEditor () {
+    setEditorValue(shootItem.value)
+  }
+
+  function reloadEditor () {
+    resetEditor()
+  }
+
+  function focusEditor () {
+    cm?.focus()
+  }
+
+  function getDocumentValue () {
+    return cm?.getDocValue()
+  }
+
+  function setDocumentValue (value) {
+    cm?.setDocValue(value)
+  }
+
+  function clearDocumentHistory () {
+    cm?.clearDocHistory()
```

Comment:
call dispatchHistoryEffect directly

Reply:
I decided to keep the method `cm.clearDocHistory` because it is used twice, and in this particular case, the following is necessary according to @grolu.
```
cm.dispatchHistoryEffect(false)
cm.dispatchHistoryEffect(true)
```
Encapsulating this logic in a function improves code readability, prevents duplication, and reduces the risk of errors if the same logic needs to be applied in multiple places.